pkg/engine/lifecycletest/resource_hooks_test.go

```diff
@@ -0,0 +1,394 @@
+
+		_, err = monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			ResourceHookBindings: deploytest.ResourceHookBindings{
+				BeforeCreate: []*deploytest.ResourceHook{myHook},
+			},
+		})
+		require.ErrorContains(t, err, "resource monitor shut down while waiting on step's done channel")
+		return nil
+	})
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+	}
+	p.Steps = []lt.TestStep{{
+		Op:            Update,
+		SkipPreview:   true,
+		ExpectFailure: true,
+		Validate: func(project workspace.Project, target deploy.Target, entries JournalEntries,
+			evts []Event, err error,
+		) error {
+			sawFailure := false
+			for _, evt := range evts {
+				if evt.Type == DiagEvent {
+					e := evt.Payload().(DiagEventPayload)
+					sawFailure = strings.Contains(e.Message, "hook \"myHook\" failed: Oh no") &&
+						e.Severity == diag.Error && e.URN.Name() == "resA"
+					if sawFailure {
+						break
+					}
+				}
+			}
+
+			assert.True(t, sawFailure, "There should be an error diagnostic for `resAB`")
+			return err
+		},
+	}}
+	snap := p.Run(t, nil)
+	assert.True(t, hookCalled)
+	assert.Len(t, snap.Resources, 1)
+}
+
+func TestResourceHookAfterDelete(t *testing.T) {
+	t.Parallel()
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+	}
+
+	createResource := true
+	hookCalled := false
+
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		callbacks, err := deploytest.NewCallbacksServer()
+		require.NoError(t, err)
+		defer func() { require.NoError(t, callbacks.Close()) }()
+
+		fun := func(ctx context.Context, urn resource.URN, id resource.ID, inputs resource.PropertyMap,
+			outputs resource.PropertyMap,
+		) error {
+			hookCalled = true
+			assert.Equal(t, urn, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"))
+			return nil
+		}
+		myHook, err := deploytest.NewHook(monitor, callbacks, "myHook", fun, true)
+		require.NoError(t, err)
+
+		if createResource {
+			_, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+				ResourceHookBindings: deploytest.ResourceHookBindings{
+					AfterDelete: []*deploytest.ResourceHook{myHook},
+				},
+			})
+			assert.NoError(t, err)
+		}
+
+		err = monitor.SignalAndWaitForShutdown(context.Background())
+		assert.NoError(t, err)
+		return nil
+	})
+
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+	}
+	project := p.GetProject()
+
+	// Run an update to create the resource
+	snap, err := lt.TestOp(Update).RunStep(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, nil, "0")
+	require.NoError(t, err)
+	assert.NotNil(t, snap)
```

Comment:
```suggestion
	require.NotNil(t, snap)
```

Subsequent steps will panic if `snap` is nil.